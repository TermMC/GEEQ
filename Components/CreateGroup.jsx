import { useNavigation } from '@react-navigation/core';
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View, TextInput, Button } from 'react-native';
import { Link } from '@react-navigation/native';
import {
  collection,
  doc,
  addDoc,
  getDocs,
  getDoc,
  updateDoc,
  arrayUnion,
  setDoc,
  query,
  where,
  collectionGroup,
  document,
  serverTimestamp,
  Timestamp,
} from 'firebase/firestore';
import { auth, db } from '../firebase';

import { useState, useEffect } from 'react';
import GlobalCSS from '../GlobalCSS';
import { ScrollView } from 'react-native-web';
import ViewGroups from './ViewGroups';

export default function CreateGroup({ navigation }) {
  //const navigation = useNavigation();
  const collRef = collection(db, 'groups');
  const [user, setUser] = useState({});
  const uid = auth.currentUser.uid; //ef83N7qN5beB5oJtm9CgCnW7Ydv1
  console.log('auth ', uid);

  const docRef = doc(db, 'users', uid);

  const [groupName, setGroupName] = useState('');
  const [groupAvatar, setGroupAvatar] = useState(''); // https://freesvg.org/img/group.png

  // get a single user
  const getUser = async () => {
    const udocs = await getDoc(docRef);
    setUser({ uid: udocs.id, ...udocs.data() });

    console.log(' getSingleDoc ', udocs.id, udocs.data());
  };

  useEffect(() => {
    getUser();
  }, []);

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (groupName.trim() === '') {
      alert('Please enter new group name');
      return;
    }

    let timestamp = Timestamp.now();

    // add to groups collection
    const newGroupCreated = await addDoc(collRef, {
      group_name: groupName,
      avatar: groupAvatar,
      created_at: timestamp,
      users: [{ name: user.name, uid }],
    });

    // add this group to users doc
    console.log('new group id: ', newGroupCreated.id);
    const newGroup = {
      group_name: groupName,
      group_id: newGroupCreated.id,
      created_at: timestamp,
    };

    const docRef = doc(db, 'users', uid);
    updateDoc(docRef, { groups: arrayUnion(newGroup) });

    console.log('form sumitted');
  };

  return (
    <>
      <View style={GlobalCSS.container}>
        <Text>Create a new group</Text>

        <View style={{ padding: 10 }}>
          <TextInput
            style={{ height: 40 }}
            placeholder="Group name!"
            onChangeText={(newText) => setGroupName(newText)}
            defaultValue={groupName}
          />
          <TextInput
            style={{ height: 40 }}
            placeholder="Avatar url"
            onChangeText={(newText) => setGroupAvatar(newText)}
            defaultValue={groupAvatar}
          />

          <Button title="Submit" onPress={handleSubmit} />
        </View>
      </View>
      <Link
        to={{
          screen: 'ViewGroups',
        }}
      >
        View all groups
      </Link>

      <Link
        to={{
          screen: 'Group',
          params: { group_id: '4cXw12VSrQoKHmKsL1Di' },
        }}
      >
        Go to a group
      </Link>
    </>
  );
}
