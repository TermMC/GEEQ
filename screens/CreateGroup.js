import { Text, View, TextInput, Button, Image } from "react-native";
import {
  collection,
  doc,
  addDoc,
  getDoc,
  updateDoc,
  arrayUnion,
  Timestamp,
} from "firebase/firestore";
import { auth, db } from "../firebase";
import { useState, useEffect, useContext } from "react";
import GlobalCSS from "../GlobalCSS";
import { ScrollView } from "react-native-web";
import { ThemeContext } from "../Context/Context";
import InviteTest from "../Components/InviteTest";

export default function CreateGroup({ navigation }) {
  const theme = useContext(ThemeContext);
  const collRef = collection(db, "groups");
  const [user, setUser] = useState({});
  const uid = auth.currentUser.uid;
  console.log("auth ", uid);

  const docRef = doc(db, "users", uid);

  const [groupName, setGroupName] = useState("");
  const [groupAvatar, setGroupAvatar] = useState("");

  // get a single user
  const getUser = async () => {
    const udocs = await getDoc(docRef);
    setUser({ uid: udocs.id, ...udocs.data() });

    console.log(" getSingleDoc ", udocs.id, udocs.data());
  };

  useEffect(() => {
    getUser();
  }, []);

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (groupName.trim() === "") {
      alert("Please enter new group name");
      return;
    }

    let timestamp = Timestamp.now();

    // add to groups collection
    const newGroupCreated = await addDoc(collRef, {
      group_name: groupName,
      avatar: groupAvatar,
      created_at: timestamp,
      users: [{ name: user.name, uid }],
    });

    // add this group to users doc
    console.log("new group id: ", newGroupCreated.id);
    const newGroup = {
      group_name: groupName,
      group_id: newGroupCreated.id,
      created_at: timestamp,
    };

    const docRef = doc(db, "users", uid);
    updateDoc(docRef, { groups: arrayUnion(newGroup) }).then(
      navigation.navigate("Group", { group_id: newGroupCreated.id })
    );

    console.log("form sumitted");
  };

  return (
    <>
      <View style={theme.homeContainer}>
        <Text style={theme.header}>Create a new group</Text>

        <View style={theme.container}>
          <TextInput
            style={theme.header2}
            placeholder="Group name!"
            onChangeText={(newText) => setGroupName(newText)}
            defaultValue={groupName}
          />
          <TextInput
            style={theme.header2}
            placeholder="Avatar url"
            onChangeText={(newText) => setGroupAvatar(newText)}
            defaultValue={groupAvatar}
          />
          {groupAvatar ? (
            <Image
              source={{ uri: groupAvatar }}
              style={{
                height: 200,
                width: 200,
                borderRadius: "100%",
              }}
            />
          ) : (
            <Text style={theme.header3}>
              No Avatar? No Problem! <br /> Your group will be assigned a random
              image from the vaults.
            </Text>
          )}
          <Button style={theme.button} title="Submit" onPress={handleSubmit} />
        </View>
      </View>

      {/* <Link
        to={{
          screen: 'ViewGroups',
        }}
      >
        View all groups
      </Link>

      <Link
        to={{
          screen: 'Group',
          params: { group_id: '4cXw12VSrQoKHmKsL1Di' },
        }}
      >
        Go to a group
      </Link> */}
    </>
  );
}
