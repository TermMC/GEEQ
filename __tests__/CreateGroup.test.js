// import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import CreateGroup from '../Components/CreateGroup';
import Intro from '../Intro.jsx';

test('Cannot render without login', () => {
  const tree = renderer.create(<CreateGroup />).toJSON();
  expect(tree).toMatchSnapshot();
});

it(' screen', () => {
  const page = render(<Intro />);

  const texts = page.queryAllByText('Add');
});
