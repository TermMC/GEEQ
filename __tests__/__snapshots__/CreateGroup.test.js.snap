// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cannot render without login first 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text>
      Create a new group
    </Text>
    <View
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        defaultValue=""
        onChangeText={[Function]}
        placeholder="Group name!"
        rejectResponderTermination={true}
        style={
          Object {
            "height": 40,
          }
        }
        underlineColorAndroid="transparent"
      />
      <TextInput
        allowFontScaling={true}
        defaultValue=""
        onChangeText={[Function]}
        placeholder="Avatar url"
        rejectResponderTermination={true}
        style={
          Object {
            "height": 40,
          }
        }
        underlineColorAndroid="transparent"
      />
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <Text
    accessibilityRole="link"
    href="/ViewGroups"
    onPress={[Function]}
  >
    View all groups
  </Text>,
  <Text
    accessibilityRole="link"
    href="/Group?group_id=4cXw12VSrQoKHmKsL1Di"
    onPress={[Function]}
  >
    Go to a group
  </Text>,
]
`;
